### ISSUE TITLE
Snap to grid feature?

### ISSUE THREAD

**Issue by ariel-phet** [phetsims/charges-and-fields#98]
@arouinfar 

From an unfuddle issue:

> Currently the location of the electric field vectors are not on the same coordinate system as "grid".
> This has the unfortunate effect of making a point charge not look the textbooks with the field lines being symmetric and radially outward. (see attached)  They are correct, but visually are a bit disconcerting.
> 
> It would be great if the spacing were 0.25 m for the location of the electric field vectors and coincident with the grid axes.  This way if you chose to put a charge on a grid intersection the field would be clearer.

Unfortunately, I don't think I had reviewed the unfuddle tickets for this sim (my miss, and might have been before we had a solid process in place)

I don't think the suggestion above is a good idea (I can see why Mike chose the arrangement he did - if you place a charge at the major intersections of the grid, the field points radially outward as shown below)

![point charge](https://cloud.githubusercontent.com/assets/4051174/18064304/0c205dac-6dec-11e6-8e9f-cf2a7c0ef90f.PNG)

However, we could possibly consider a snap to grid feature (a checkbox that can be activated when you activate the grid) We would definitely not want such snapping to grid to be default.

Let me know if you think this might be a useful feature. If so I might discuss with @andrewadare on how long it would be to implement (but likely would go into a V2)

**Comment by arouinfar** [comment]
> I don't think the suggestion above is a good idea

I agree. Earlier in the design process there was some discussions about the locations of the field arrows within the grid. In Flash, the E-field arrows were located on grid lines, but the major intersections didn't hold much importance.
![image](https://cloud.githubusercontent.com/assets/8419308/18065137/cff061e8-6def-11e6-84fd-c2ed26ea8d8c.png)

We deliberately shifted the grid a bit so that the charges placed on the intersections of major gridlines would show the radial pattern in @ariel-phet's screenshot. We also spaced the E-field arrows out so that they would have consistent placement on the grid. These changes satisfied a user suggestion from #25.

> we could possibly consider a snap to grid feature (a checkbox that can be activated when you activate the grid) We would definitely not want such snapping to grid to be default.

I think this would be a nice feature to consider for v2, and agree that we'd want it off by default. I could see this being particularly useful when it comes to keyboard navigation.

**Comment by andrewadare** [comment]
I don't (yet) know how to implement this, but I'm sure it's just a matter of pinging the more experienced developers to point out some comparable examples. I assume this should not hold up RC1, but we can leave this issue open for version 2.

**Comment by arouinfar** [comment]
@andrewadare I don't think @ariel-phet meant rc.2 - I certainly didn't. I don't imagine we want this to hold back publication of the sim, but this feature could be in a future version.

**Comment by ariel-phet** [comment]
@andrewadare correct, this would be a feature to consider _after_ the sim is published, for a new release at some point

**Comment by ariel-phet** [comment]
Unassigning for the moment

**Comment by andrewadare** [comment]
On hold; reconsider for Charges and Fields 2.0.

**Comment by veillette** [comment]
This feature is relatively easy to implement. 
Should the feature exists only for the charges or applies to the electric sensors and the electricPotential tool  as well?

**Comment by arouinfar** [comment]
@veillette I think we would want to add a nested checkbox beneath Grid -- not sure what we would call it. When on, the charges and sensors would snap to the grid intersections. I hadn't thought about locking the electricPotential tool to the grid, but it may also be nice to snap its crosshairs to the grid, as well. What do you think @ariel-phet?

**Comment by ariel-phet** [comment]
@arouinfar @veillette I think if we make a "snap to grid" it would make sense to snap everything to the grid while that feature was active. It might be handy for making certain charge arrangements (especially symmetric ones)

**Comment by ariel-phet** [comment]
Since @veillette says the feature is easy to implement, putting it at low priority if he desires to work on it.

**Comment by veillette** [comment]
In the process of implementing this feature, there are additional questions that crop up. I have not committed anything yet.

 Here is the panel
![image](https://user-images.githubusercontent.com/7633344/27975017-7de2a976-632e-11e7-9892-5cefbe3e166d.png)

I've put a placeholder string until we find an appropriate value.


My first question is the behavior if we checked off grid but snapped to grid is still checked on. 
![image](https://user-images.githubusercontent.com/7633344/27974973-4f497e32-632e-11e7-8fc8-dcc83e5ad834.png)

-Should the snap to grid behavior be disabled in this case? 

My second question is the behavior of the charges and other components when the client is checking the snap to grid feature. 

- Should the already present charges and sensors be snapped to the grid at that instant or is the snapped feature affecting only the position of the components at the end of a drag event.

My third question is regarding which grid I should used to snap on the charge. Your picture of a charge is snapped to the major gridlines. 

- Should we snapped to be the major gridlines or the major/minor gridlines?   

Not to bias you or anything, here are my choices.
(1) disabled
(2) snapped to grid all previous charges
(3) user minor and major

I asked a physics faculty to test drive the simulation. His answers were
(1) disabled
(2) snapped to grid only after drag event
(3) use minor and major 

He also liked to that the sensors can be snapped to the grid, although personally after using the simulation for a while, I would only snapped the charges to the grid. 

Assigning to @arouinfar for your input.

**Comment by veillette** [comment]
One more question, 
should we snap the measuring tape as well ?

**Comment by arouinfar** [comment]
> I've put a placeholder string until we find an appropriate value.

@veillette I think "Snap to Grid" is straightforward, and I haven't thought of anything cleverer, so let's go with that for now. To be a bit nitpicky, I would recommend that we consistently use title casing, so "to" should be lowercase, and "Only" in "Direction only" should be capitalized.


To answer your questions:

(1) Disabled, but still visibly checked/grayed out. I think the behavior should mirror the Electric Field/Direction Only relationship. If Electric Field is unchecked, Direction Only can remain checked -- but it's moot. There is no visible electric field, so nothing to show the direction of. The same could be argued for Grid/Snap to Grid. When Grid is off, there is no grid to snap to. The nested checkbox can still be grayed out and checked, but technically the feature is turned off until Grid is turned back on.

(2) I think charges already in the play area should snap to the nearest grid intersection when Snap to Grid is activated.

(3) Snap to any grid intersection, major or minor

> He also liked to that the sensors can be snapped to the grid, although personally after using the simulation for a while, I would only snapped the charges to the grid.

I could make an argument either way. On one hand, snapping the sensors to the grid allows for a more precise measurement of the field. On the other hand, it could be difficult to show that the E-field is perpendicular to the equipotential curve if the sensors must be snapped to the grid. However, this could be demonstrated easily with Snap to Grid turned off, so I'm still in favor of snapping the sensors to the grid.

> Should we snap the measuring tape as well ?

Great question. If we snap the crosshairs of the equipotential tool to the grid, then we should do the same with the measuring tape. Snapping the equipotential tool to the grid would ensure the equipotential curve is being drawn at a precise distance from the charges (for that one point, at least), which could be helpful. 

@veillette I'd also like to point out that there were several issues related to overlapping charges or discontinuities that cropped up in QA testing (such as #46, #82, #84, #85, #88, #89). While these issues were all addressed before publication, it's possible that snapping things to the exact same location may cause some related side effects.

**Comment by veillette** [comment]
Thanks @arouinfar for pointing out all the previous issues.  The snappingToGrid feature needs a model that handles well infinity. Scenarios that used to be corner cases will  be much more prevalent.

**Comment by veillette** [comment]
@arouinfar, I commited the changes according to your request. I created a separate issue #126  on how we should handle charges that are overlapping since there is a much higher likelihood of having multiple overlapping charges with the snap to grid features.

In addition, we will need someone to review this commit since it involves PhET-IO which i was not able to able to test.

However, this should not prevent you from testing out the features on master. Let me know if it is working as expected or if it needs any adjustments.

**Comment by arouinfar** [comment]
@veillette I really like the snap to grid feature! It's easy to use, and I love being able to create symmetric arrangements. 

I ran into one situation where the charge was delayed in snapping to the grid. To reproduce:

1. Turn on Grid, then purposely place a charge so that it is misaligned to the grid
2. Turn on Snap to Grid, the charge snaps to the nearest grid intersection (as it should)
3. Turn off Grid (Snap to Grid is checked but grayed out) and move the charge
4. Turn on Grid, charge is misaligned to grid but doesn't snap
5. Click on charge, it snaps to the grid

**Comment by arouinfar** [comment]
@veillette I also like that the equipotential tool displays infinity when snapped on top of a charge, but I would prefer to use the symbol, rather than the word.

**Comment by veillette** [comment]
The display infinity is being tracked in a separate issue ( #127)

**Comment by oliver-phet** [comment]
This was requested by another user on March 7, 2018:

> I created a lab using the charges and fields simulation, where the students determine the value of k. In talking about errors in measurement, several students commented that it would be nice if there was an option for “snap to grid”. I thought I would send you a note to put the idea out there.

**Comment by arouinfar** [comment]
@ariel-phet do you think there would be any bandwidth to deal with this issue any time soon?

**Comment by arouinfar** [comment]
I believe everything has been implemented (except for #127) and the changes still need to go through QA testing.

**Comment by ariel-phet** [comment]
@arouinfar let me ask @veillette - he has been dedicating a bit of time to PhET this semester with mentoring students, and he might have time to polish this off.

**Comment by ariel-phet** [comment]
@veillette said:

> Sure, I'd be happy to. I'll do it over the weekend.
> Martin

**Comment by oliver-phet** [comment]
In ES6 migration testing, I noticed using the Reset All button does not uncheck "Snap to Grid".

**Comment by oliver-phet** [comment]
Another user emailed phethelp with this request. I'll let them know when this is published.

> the size of the charges makes it difficult to place them it would be nice if there was an option to “lock” them into a specific grid position
> this would also help in measure the distances with the tape measure, along with a “fine tuning” ability to better control the placement of the tape measure and the distance increments. My “mouse control” is also a tad bit problematic.

**Comment by oliver-phet** [comment]
Another email:
> I would like a "Snap to Grid" option for placing charges.
> It would help me make lines of charge.
